-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-556b13/40121/summary.html
**COMMIT**: 250cec224395db2865d088ae39b9af07ee6a022b
**CMSSW**: CMSSW_14_1_X_2024-06-27-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/45327/40121/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling  src/Utilities/StorageFactory/src/Throw.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH='CMSSW_14_1_X_2024-06-27-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_1_X_2024-06-27-1100' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-06-27-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/pcre/8.43-e34796d17981e9b6d174328c69446455/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/boost/1.80.0-a1544032d9d65904ac2112b6d35bba55/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/lcg/root/6.30.07-5b5d83179d6ed7bc60aa7a5006899eba/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/tbb/v2021.9.0-1dc6d66b4b014f4ae4733b04914ce7d4/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/zlib/1.2.11-1a082fc322b0051b504cc023f21df178/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/fmt/8.0.1-258b4791803c34b7e98cf43693e54d87/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/tinyxml2/6.2.0-88fe0ec301baf763fa3c485e5b67ed91/include -D_FILE_OFFSET_BITS=64 -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++17 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v2 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/Utilities/StorageFactory/src/UtilitiesStorageFactory/Throw.cc.d src/Utilities/StorageFactory/src/Throw.cc -o tmp/el8_amd64_gcc12/src/Utilities/StorageFactory/src/UtilitiesStorageFactory/Throw.cc.o
>> Compiling  src/Utilities/StorageFactory/src/UnixFile.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH='CMSSW_14_1_X_2024-06-27-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_1_X_2024-06-27-1100' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-06-27-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/pcre/8.43-e34796d17981e9b6d174328c69446455/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/boost/1.80.0-a1544032d9d65904ac2112b6d35bba55/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/lcg/root/6.30.07-5b5d83179d6ed7bc60aa7a5006899eba/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/tbb/v2021.9.0-1dc6d66b4b014f4ae4733b04914ce7d4/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/zlib/1.2.11-1a082fc322b0051b504cc023f21df178/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/fmt/8.0.1-258b4791803c34b7e98cf43693e54d87/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02843/el8_amd64_gcc12/external/tinyxml2/6.2.0-88fe0ec301baf763fa3c485e5b67ed91/include -D_FILE_OFFSET_BITS=64 -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++17 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v2 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/Utilities/StorageFactory/src/UtilitiesStorageFactory/UnixFile.cc.d src/Utilities/StorageFactory/src/UnixFile.cc -o tmp/el8_amd64_gcc12/src/Utilities/StorageFactory/src/UtilitiesStorageFactory/UnixFile.cc.o
src/Utilities/StorageFactory/src/RemoteFile.cc: In static member function 'static int edm::storage::RemoteFile::local(const std::string&, std::string&)':
src/Utilities/StorageFactory/src/RemoteFile.cc:57:5: error: 'fs' has not been declared
   57 |     fs::path current_path = fs::current_path();
      |     ^~
src/Utilities/StorageFactory/src/RemoteFile.cc:58:12: error: 'current_path' was not declared in this scope; did you mean 'std::filesystem::current_path'?
   58 |     temp = current_path.string();
      |            ^~~~~~~~~~~~
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Local Products Rules ..... started
>> Local Products Rules ..... done
>> Creating project symlinks
>> Entering Package Utilities/StorageFactory
>> Compile sequence completed for CMSSW CMSSW_14_1_X_2024-06-27-1100
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
+ eval scram build outputlog '&&' '(python3' /data/cmsbld/jenkins/workspace/ib-run-pr-tests/cms-bot/buildLogAnalyzer.py --logDir /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_1_X_2024-06-27-1100/tmp/el8_amd64_gcc12/cache/log/src '||' 'true)'
++ scram build outputlog
>> Entering Package Utilities/StorageFactory
>> Compiling  src/Utilities/StorageFactory/bin/anycp.cpp
Entering library rule at src/Utilities/StorageFactory/plugins
</pre>

--- a/Utilities/StorageFactory/src/RemoteFile.cc
+++ b/Utilities/StorageFactory/src/RemoteFile.cc
@@ -10,6 +10,7 @@
 #include <ostream>
 #include <cstring>
 #include <filesystem>
+#include <vector>
 #if __APPLE__
 #include <crt_externs.h>
 #define environ (*_NSGetEnviron())
@@ -53,28 +54,31 @@ int RemoteFile::local(const std::string &tmpdir, std::string &temp) {
   // This is better for grid jobs as the current directory is
   // likely to have more space, and is more optimised for
   // large files, and is cleaned up after the job.
+  std::vector<char> temp_chars;
+
   if (tmpdir.empty() || tmpdir == ".") {
-    fs::path current_path = fs::current_path();
-    temp = current_path.string();
+    std::filesystem::path current_path = std::filesystem::current_path();
+    temp_chars.assign(current_path.string().begin(), current_path.string().end());
   } else {
-    temp = tmpdir;
+    temp_chars.assign(tmpdir.begin(), tmpdir.end());
   }
 
-  if (temp.back() != '/')
-    temp += '/';
+  if (temp_chars.back() != '/')
+    temp_chars.push_back('/');
 
-  temp += "storage-factory-local-XXXXXX";
+  std::string suffix = "storage-factory-local-XXXXXX";
+  temp_chars.insert(temp_chars.end(), suffix.begin(), suffix.end());
 
-  // Make the string null-terminated for mkstemp
-  std::vector<char> temp_chars(temp.begin(), temp.end());
+  // Ensure the vector is null-terminated
   temp_chars.push_back('\0');
 
   int fd = mkstemp(temp_chars.data());
-
   if (fd == -1)
-    throwStorageError("RemoteFile", "Calling RemoteFile::local()", "mkstemp()", errno);
+    throw std::system_error(errno, std::generic_category(), "mkstemp failed");
 
-  temp = std::string(temp_chars.data());
+  // Copy temp_chars to temp
+  temp.assign(temp_chars.begin(), temp_chars.end() - 1);  // Exclude the null terminator
+
   return fd;
 }
 

--- a/Utilities/StorageFactory/src/RemoteFile.cc
+++ b/Utilities/StorageFactory/src/RemoteFile.cc
@@ -58,8 +58,11 @@ int RemoteFile::local(const std::string &tmpdir, std::string &temp) {
 
   if (tmpdir.empty() || tmpdir == ".") {
     std::filesystem::path current_path = std::filesystem::current_path();
-    temp_chars.assign(current_path.string().begin(), current_path.string().end());
+    auto spath = current_path.string();
+    temp_chars.reserve(spath.size() + 30);
+    temp_chars.assign(spath.begin(), spath.end());
   } else {
+    temp_chars.reserve(tmpdir.size() + 30);
     temp_chars.assign(tmpdir.begin(), tmpdir.end());
   }
 
@@ -74,8 +77,7 @@ int RemoteFile::local(const std::string &tmpdir, std::string &temp) {
 
   int fd = mkstemp(temp_chars.data());
   if (fd == -1)
-    throw std::system_error(errno, std::generic_category(), "mkstemp failed");
-
+    throwStorageError("RemoteFile", "Calling RemoteFile::local()", "mkstemp()", errno);
   // Copy temp_chars to temp
   temp.assign(temp_chars.begin(), temp_chars.end() - 1);  // Exclude the null terminator
 

